### ISSUE TITLE
Inquiry on Help to Reproduce the Paper with Newer LLMs

### ISSUE THREAD

**Issue by cs329yangzhong** [dongqi-me/RST_LoRA#1]
Hi @dongqi-me, Thanks for open-sourcing the code for your RST-Lora paper! I am currently working on reproducing the experiments from your paper, particularly focusing on implementing the improved LoRA model and applying them to newer LLMs such as  Qwen2.5 and Gemma3 m.odels.  However, I have encountered some challenges due to the absence of documentation or a README file detailing the specific modifications made to the standard LoRA modules in your codebase. I would greatly appreciate it if you could provide insights or documentation regarding these changes.

Additionally, I am interested in understanding how the codebase retrieves relation predictions between different Elementary Discourse Units (EDUs) using the DMRST framework. How are these predictions incorporated into the extra RST layer, as illustrated in Figure 3 of your paper? It would be immensely helpful if you could clarify which parts of the codebase handle these processes.

Thanks！

**Comment by dongqi-me** [comment]
Hi,

Thanks for your interest in our RST-LoRA work!

To address your questions:

### 📌 1. Modifications to Standard LoRA Modules

In our implementation, we extend the standard LoRA module by incorporating an additional **RST tensor**, which allows the LoRA adaptation to be modulated by discourse-aware information. Specifically, we apply this RST tensor to reweight the input embeddings before computing the low-rank updates.

Here is the relevant code snippet:

```python
import torch
import torch.nn as nn
import math

class RSTLoRA(nn.Module):
    def __init__(self, pretrained_model, rank, alpha):
        super().__init__()
        self.pretrained_model = pretrained_model  # Pretrained weights W
        
        # Initialize LoRA parameter matrices A and B
        self.lora_A = nn.Parameter(torch.zeros((rank, pretrained_model.weight.shape[1])))
        self.lora_B = nn.Parameter(torch.zeros((pretrained_model.weight.shape[0], rank)))
        
        # Initialize RST vector
        self.rst = nn.Parameter(torch.zeros(pretrained_model.weight.shape[1]))
        
        self.alpha = alpha
        self.rank = rank
        
        # Parameter initialization
        nn.init.kaiming_uniform_(self.lora_A, a=math.sqrt(5))
        nn.init.zeros_(self.lora_B)
        nn.init.normal_(self.rst, std=0.02)
    
    def forward(self, x):
        # RST operation: copy the input x and apply RST vector
        x_copy = x.clone()
        
        # LoRA computation: low-rank matrix multiplication BA
        lora_output = (self.lora_B @ self.lora_A) * (self.alpha / self.rank)
        last_output = self.pretrained_model(x)


        # Combined output
        output = last_output + (x_copy * (1 + self.rst)) @ lora_output
        
        return output
```

### 📌 2. Integration of DMRST Relation Predictions

As for the integration of RST information:

- We **extract the logits** from the **final layer of the DMRST model**, which represent the predicted rhetorical relations between EDUs.
- These logits are processed into a fixed-length **RST tensor**, which is used as the vector `self.rst` in the `RSTLoRA` class above.
